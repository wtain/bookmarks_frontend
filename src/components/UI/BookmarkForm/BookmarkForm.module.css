.bookmark_form {
    width: 300px;
    padding: 0 0 15px 15px;
    margin: 0px;
    border-radius: 16px;
    background-color: darkslategray;
    border: solid 2px;
}

.form_valid {
    border-color: black;
}

.form_invalid {
    border-color: red;
}

.input {
    width: 280px;
    border: 1px solid teal;
    background: transparent;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    margin: 5px 0 5px;
    color: white;
}

.textarea_wrapper {
    width: 280px;
}

.tags_container {
    width: 278px;
}

.textarea {
    /* width: calc(100%-30px); */
    /* max-width: 100%; */
    width: 280px;
    min-width: 280px;
    max-width: 280px;
    min-height: 5ex;
    max-height: 30ex;
    background: transparent;
    border: 1px solid teal;
    /* margin: 5px 5px 5px 5px; */
    /* padding: 5px 5px 5px 5px; */
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    margin: 5px 0 5px;
    height: 10ex;
    color: white;
}

.button {
    color: teal;
    font-size: 14px;
    background: transparent;
    border: 1px solid teal;
    cursor: pointer;
}
