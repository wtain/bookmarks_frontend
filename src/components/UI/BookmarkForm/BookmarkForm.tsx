
import React, { useEffect, useRef, useState } from "react";
import BookmarkDto from "../../../domain/dto/BookmarkDto";
import cl from './BookmarkForm.module.css'
import TagsEditor from "../TagsEditor/TagsEditor";
import TagDto from "../../../domain/dto/TagDto";
import BookmarkUtils from "../../../utils/BookmarkUtils";

interface Props {
    create: (bookmark: BookmarkDto) => void;
    visibility: boolean;
    validation: (bookmark: BookmarkDto) => boolean;
}

const BookmarkForm: React.FC<Props> = (props: Props) => {

    const EmptyBookmark = { summary: "", contents: "" };

    const [bookmark, setBookmark] = React.useState(EmptyBookmark);
    const inputRef = useRef<HTMLInputElement | null>(null);
    const [tags, setTags] = useState<TagDto[]>([])
    
    const [valid, setValid] = React.useState(false);
    
    useEffect(() => {
        if (props.visibility) {
            if (inputRef.current !== null) {
                inputRef.current.focus();
            }
        }
        setBookmark(EmptyBookmark);
        setTags([]);
    }, [props.visibility])

    const updateBookmark = (bookmark: BookmarkDto) => {
        setBookmark(bookmark);
        setValid(props.validation(bookmark));
    }

    const doAddBookmark = () => {
        const newBookmark: BookmarkDto = BookmarkUtils.createBookmark(bookmark.summary, bookmark.contents, tags);
        if (!valid) {
            return;
        }
        props.create(newBookmark);
        setBookmark(EmptyBookmark);
        setValid(false);
    }

    const addBookmark = (e: React.MouseEvent) => {
        e.preventDefault();
        doAddBookmark();
    }

    const form_classes = [cl.bookmark_form];
    form_classes.push(valid ? cl.form_valid : cl.form_invalid);

    return (
        <form className={form_classes.join(" ")}>
            <input
                    placeholder="Summary" 
                    value={bookmark.summary} 
                    onChange={(e) => updateBookmark({...bookmark, summary: e.target.value} as BookmarkDto)} 
                    className={cl.input}
                    ref={inputRef}
                    />
            <div className={cl.textarea_wrapper}>
                <textarea rows={3}
                            placeholder="Contents" 
                            value={bookmark.contents} 
                            onChange={(e) => updateBookmark({...bookmark, contents: e.target.value} as BookmarkDto)} 
                            className={cl.textarea}
                            onKeyDown={(e) => {
                                if (e.ctrlKey && e.code === "Enter") {
                                   doAddBookmark();
                                   e.stopPropagation();
                                }
                            }}
                            />
            </div>
            <TagsEditor 
                    containerClass={cl.tags_container}
                    tags={tags} 
                    onTagAdded={(tag: TagDto) => {
                        setTags([...tags, tag]);
                    }} 
                    onDelete={(index: number) => {
                        const new_tags = tags.filter((v, i) => i !== index);
                        setTags(new_tags)
                    }} />
            <button onClick={addBookmark} className={cl.button}>Add</button>
        </form>
    )
}

export default BookmarkForm;