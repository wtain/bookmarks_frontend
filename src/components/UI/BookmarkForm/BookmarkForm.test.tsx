import renderer from 'react-test-renderer';
import BookmarkForm from './BookmarkForm';

function run_test(visibility: boolean) {
  const tree = renderer
  .create(
    <BookmarkForm visibility={visibility}  />
  )
  .toJSON();
  expect(tree).toMatchSnapshot();
}

it('CalendarPage renders correctly visible', () => {
  run_test(true)
});

it('CalendarPage renders correctly hidden', () => {
  run_test(false)
});