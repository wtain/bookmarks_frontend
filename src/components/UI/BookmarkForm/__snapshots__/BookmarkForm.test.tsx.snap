// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarPage renders correctly hidden 1`] = `
<form
  className="bookmark_form form_invalid"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Summary"
    value=""
  />
  <div
    className="textarea_wrapper"
  >
    <textarea
      className="textarea"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Contents"
      rows={3}
      value=""
    />
  </div>
  <div
    className="container tags_container"
    onClick={[Function]}
  >
    <span
      className="sp"
    >
      <input
        className="input"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Tags"
        value=""
      />
    </span>
  </div>
  <button
    className="button"
    onClick={[Function]}
  >
    Add
  </button>
</form>
`;

exports[`CalendarPage renders correctly visible 1`] = `
<form
  className="bookmark_form form_invalid"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Summary"
    value=""
  />
  <div
    className="textarea_wrapper"
  >
    <textarea
      className="textarea"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Contents"
      rows={3}
      value=""
    />
  </div>
  <div
    className="container tags_container"
    onClick={[Function]}
  >
    <span
      className="sp"
    >
      <input
        className="input"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Tags"
        value=""
      />
    </span>
  </div>
  <button
    className="button"
    onClick={[Function]}
  >
    Add
  </button>
</form>
`;
